### PR TITLE
Correctly clean thread from the waiting list if it was blocked and ki…

### PR DESCRIPTION
**Pull request by jxyang**
…lled

Signed-off-by: Xuejun Yang <xuejya@microsoft.com>

### CODE REVIEW

File: kernel/thread.c [hunk 1]
```diff
@@ -237,6 +237,28 @@ static void _free_zombies(void* arg)
 
 void myst_zombify_thread(myst_thread_t* thread)
 {
+    // If the thread was blocked on a condvar and got killed,
+    // get out of the waiting list of the condvar.
+    // If the thread exits normally, the cond_wait should be NULL.
```
Comment (Francis-Liu) [inline kernel/thread.c:242]: Can a thread be killed but was not blocked on a condvar?

File: kernel/thread.c [hunk 2]
```diff
@@ -237,6 +237,28 @@ static void _free_zombies(void* arg)
...
+    myst_cond_signal_thread(thread->signal.cond_wait, thread);
```
Comment (Francis-Liu) [inline kernel/thread.c:243]: curious: why not check `cond_wait != NULL` before calling? To me, EINVAL makes me feel it's an error.

File: kernel/cond.c [hunk 3]
```diff
@@ -216,6 +252,7 @@ int myst_cond_requeue(
         for (myst_thread_t* p = requeues.front; p; p = next)
         {
             next = p->qnext;
+            p->signal.cond_wait = c2;
```
Comment (Francis-Liu) [inline kernel/cond.c:255]: Before your fix, requeued threads had their `cond_wait=NULL`?

File: kernel/cond.c [hunk 4]
```diff
@@ -64,6 +65,14 @@ int myst_cond_timedwait(
         /* Add the self thread to the end of the wait queue */
         myst_thread_queue_push_back(&c->queue, self);
 
+        // assert(self->signal.cond_wait == NULL);
```
Comment (Francis-Liu) [inline kernel/cond.c:68]: Why comment out this assert?

**Review by mikbras — APPROVED** [review]
Great work! Thanks for doing this. Approved.

Maybe there is some way to use a doubly-linked list in the future so removal of an arbitrary thread can be done with O(1) time complexity.

File: kernel/thread.c (hunk 1 above)
Comment (jxyang) [reply kernel/thread.c:242]: Yes. In that case, the field `thread->signal.cond_wait` would be NULL.

File: kernel/thread.c (hunk 2 above)
Comment (jxyang) [reply kernel/thread.c:243]: Works either way. I just prefer to keep the caller logic simpler.

File: kernel/cond.c (hunk 3 above)
Comment (jxyang) [reply kernel/cond.c:255]: The requeued threads have `cond_wait` to be `c1` without my fix.

File: kernel/cond.c (hunk 4 above)
Comment (jxyang) [reply kernel/cond.c:68]: It was failing before the final fix. I am not sure if we should enable it now.

**Pull request merged**